upgrade-takeover.sh

```diff
@@ -0,0 +1,460 @@
+#
+# $1 -- expected greater version
+# $2 -- version to compare
+function version_gt() {
+    test "$(echo "$@" | tr " " "\n" | sort -V | head -n 1)" != "$1"
+}
+
+# Compare the provided percentage and state values with backend device API
+# provisioning_progress and provisioning_state.
+#
+# $1 -- local percentage
+# $2 -- local state
+#
+# Return 0 if the values are equal; otherwise return 1
+function compare_device_state() {
+    pct=$1
+    state=$2
+    resp=$(CURL_CA_BUNDLE="${TMPCRT}" ${CURL} --header "Authorization: Bearer ${APIKEY}" \
+        "${API_ENDPOINT}/v6/device(uuid='${UUID}')?\$select=provisioning_state,provisioning_progress" | jq '.d[]')
+    remote_pct=$(echo "${resp}" | jq -r '.provisioning_progress')
+    remote_state=$(echo "${resp}" | jq -r '.provisioning_state')
+    if [ -n "${remote_pct}" ] && [ -n "${remote_state}" ]; then
+        test "${pct}" -eq "${remote_pct}" && test "${state}" = "${remote_state}"
+    else
+        return 1
+    fi
+}
+
+# Stop Supervisor and related systemd services
+function stop_services() {
+    log "Stopping supervisor and related services..."
+    systemctl stop update-balena-supervisor.timer > /dev/null 2>&1 || systemctl stop update-resin-supervisor.timer > /dev/null 2>&1
+    systemctl stop balena-supervisor  > /dev/null 2>&1 || systemctl stop resin-supervisor > /dev/null 2>&1
+    ${DOCKER_CMD} rm -f balena_supervisor resin_supervisor > /dev/null 2>&1 || true
+}
+
+# Remove and log fsck recovery files. Since takeover will reflash the disk, it's
+# not strictly necessary. However, it may indicate issues with the storage medium,
+# so good to know.
+function remove_rec_files() {
+    local boot_dir='/mnt/boot'
+    shopt -s nullglob
+    for f in "${boot_dir}"/*.REC; do
+        log WARN "Removing $f from boot partition"
+        rm -f "$f"
+    done
+    sync ${boot_dir}
+}
+
+# Download takeover tool binary for device's architecture
+# Requires $takeover_path
+# $1 -- takeover release, like v0.9.0
+# Exits script on download failure
+function download_takeover_binary() {
+    takeover_release="$1"
+    
+    architecture=$(uname -m)
+    case ${architecture} in
+        aarch64|x86_64)
+            log "Using takeover arch ${architecture}"
+            ;;
+        *)
+            log ERROR "Takeover binary for arch: ${architecture} not found"
+    esac
+
+    download_url="https://github.com/balena-os/takeover/releases/download/${takeover_release}/takeover-${architecture}-unknown-linux-musl.tar.gz"
+    log "Downloading takeover binary ${download_url}"
+
+    ${CURL} -o "${takeover_path}/takeover.tar.gz" "$download_url" || log ERROR "Could not download takeover binary, aborting."
+
+    # Extract and prepare for use
+    tar -C "${takeover_path}" -zxvf "${takeover_path}/takeover.tar.gz"
+    chown root:root "${takeover_path}/takeover"
+    rm "${takeover_path}/takeover.tar.gz"
+}
+
+# Download target balenaOS image; expects a public image
```

Comment:
I understand the API download endpoint _does_ support private images.  I am confirming with those in the know since I don't have such a device. I'll update the comment.